bin/problem.py

```diff
@@ -981,56 +981,69 @@ def validate_data(problem, mode: validate.Mode, constraints: dict | bool | None
+        samples.sort()
+        samples = [path for path, _ in zip(samples, range(3))]
+        if not samples:
+            samples = [None]
+
+        # validator, dir, read, write, copy
+        validators: list[tuple[type[validate.AnyValidator], str, str, str, list[str]]] = [
+            (validate.InputValidator, "invalid_inputs", ".in", ".in", []),
+            (validate.AnswerValidator, "invalid_answers", ".ans", ".ans", [".in"]),
+            (validate.OutputValidator, "invalid_outputs", ".ans", ".out", [".in", ".ans"]),
         ]
 
-        testcases = []
-        for cls, directory, read, write, copy, strict in validators:
-            if (p.interactive or p.multi_pass) and cls != validate.InputValidator:
-                continue
-            if not p.validators(cls, strict=True, print_warn=False):
-                continue
-            if test_path is None and copy:
-                continue
-
-            for name, data, only_strict in validator_tests.GENERATORS:
-                if not strict and only_strict:
+        testcases: list[testcase.Testcase] = []
+        for i, sample in enumerate(samples):
+            used_sample = False
+            for cls, directory, read, write, copy in validators:
+                if (p.interactive or p.multi_pass) and cls != validate.InputValidator:
+                    continue
+                if not p.validators(cls, strict=True, print_warn=False):
+                    continue
+                if any(sample is None or not sample.with_suffix(ext).exists() for ext in copy):
                     continue
 
-                if isinstance(data, str):
-                    content = data
-                elif test_path:
-                    valid = test_path.with_suffix(read).read_text()
-                    generated = data(valid)
-                    if generated is None:
+                for name, data, supported_cls in validator_tests.GENERATORS:
+                    if cls not in supported_cls:
                         continue
-                    content = generated
-
-                short_path = Path(directory) / name
-                full_path = base_path / short_path / "testcase.in"
-                full_path.parent.mkdir(parents=True, exist_ok=True)
 
-                for ext in copy:
-                    assert test_path is not None
-                    shutil.copy(test_path.with_suffix(ext), full_path.with_suffix(ext))
-                full_path.with_suffix(write).write_text(content)
-
-                testcases.append(testcase.Testcase(p, full_path, short_path=short_path))
+                    if isinstance(data, str):
+                        if i > 0 and not copy:
+                            continue
+                        content = data
+                    elif sample is None:
+                        continue
+                    elif not sample.with_suffix(read).exists():
+                        continue
```

Comment:
These two checks are based on the `sample`, not on `validator_tests.GENERATORS`, can't they be moved to an outer loop?

(note that `if sample is None` can probably be removed based on my previous comment, as well as the `assert sample is not None` on line 1034)